File: androld/app/src/main/java/z3d/base/Camera3D.java
package z3d.base;

import z3d.vo.Matrix3D;
import z3d.vo.Vector3D;

public class Camera3D extends Object3D {

    public Matrix3D camMatrix3D;
    public Matrix3D viewMatrix;
    public Matrix3D modelMatrix;
    public float distance;
    public float sceneViewHW;
    public int fovw;
    public int fovh;


    public Camera3D()
    {
        this.camMatrix3D=new Matrix3D();
        this.viewMatrix =new Matrix3D();
        this.modelMatrix =new Matrix3D();
        this.distance=100;
        this.sceneViewHW=100;
        this.fovw=300;
        this.fovh=500;
        this.upFrame();

    }
    public  void  upFrame()
    {

        this.viewMatrix.identity();
        this.viewMatrix.perspectiveFieldOfViewLH(1,1,10,1000);
        this.camMatrix3D.identity();
        this.camMatrix3D.appendRotation(this.rotationY, Vector3D.Y_AXIS);
        this.camMatrix3D.appendRotation(this.rotationX, Vector3D.X_AXIS);
        this.camMatrix3D.appendTranslation(0,0,this.distance);
        this.modelMatrix= this.viewMatrix.clone();
        this.modelMatrix.prepend(this.camMatrix3D);
        Matrix3D  m=this.camMatrix3D.clone();
        m.invert();
        Vector3D p=   m.transformVector(new Vector3D(0,0,-this.distance));
        this.x=p.x;
        this.y=p.y;
        this.z=p.z;
    }
}

File: metalios/iosgl/iosgl/pan3d/base/Camera3D.m
//
//  Camera3D.m
//  iosgl
//
//  Created by zhao on 30/12/2019.
//  Copyright © 2019 zhao. All rights reserved.
//

#import "Camera3D.h"
#import "Matrix3D.h"

@implementation Camera3D
- (instancetype)init
{
    self = [super init];
    if (self) {
        self.camMatrix3D=[[Matrix3D alloc]init];
        self.viewMatrix =[[Matrix3D alloc]init];
        self.modelMatrix =[[Matrix3D alloc]init];
        self.distance=400;
        self.sceneViewHW=100;
        self.fovw=300;
        self.fovh=300;
        [self upFrame];
    }
    return self;
}
-(void) upFrame  ;
{
  
    [self.viewMatrix identity];
    [self.viewMatrix perspectiveFieldOfViewLH:1 aspectRatio:1 zNear:10 zFar:10000];
    [self.viewMatrix prependScale:self.fovh/self.fovw y:1 z:1];
   
     
    
    [self.camMatrix3D identity];
    [self.camMatrix3D appendRotation: self.rotationY axis:Vector3D.Y_AXIS];
    [self.camMatrix3D appendRotation: self.rotationX axis:Vector3D.X_AXIS];
    [self.camMatrix3D appendTranslation: 0.0 y:0.0 z:self.distance];

    
    Matrix3D* m=[self.camMatrix3D clone];
    [m Invert];
    Vector3D* p=  [m transformVector: [[Vector3D alloc]x:0 y:0 z:-self.distance] ];
    self.x=p.x;
    self.y=p.y;
    self.z=p.z;
    self.modelMatrix= [self.viewMatrix  clone];
    [self.modelMatrix prepend:  self.camMatrix3D];
    
    
  
 
 
}
 
@end
